File: 2016/ioi_day2/messy/c/grader.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <assert.h>
#include "messy_c.h"

typedef struct node {
    struct node* zero;
    struct node* one;
} node ; 

struct node* set;
int compiled = 0;
int n;
int* p;
int w;
int r;


int read_int() {
    int x;
    assert(1 == scanf("%d", &x));
    return x;
}

// A convenience function.
int get_p(int i) {
    int ret = p[i];
    return ret;
}

struct node* create() {
    struct node* result = (struct node*)malloc(sizeof(node));
    result->zero = 0;
    result->one = 0;
    return result;
}

int main() {
    set = create();
    n = read_int();
    w = read_int();
    r = read_int();
    p = (int*)malloc(n * sizeof(int));
    for (int i = 0; i < n; i++) {
        p[i] = read_int();
    }
    int* answer = (int*)malloc(n * sizeof(int));
    restore_permutation(n, w, r, answer);
    
    
    printf("%d", answer[0]);
    for (int i = 1; i < n; i++) {
        printf(" %d", answer[i]);
    }
    printf("\n");
    return 0;
}

int check(const char* x) {
    if ((int)strlen(x) != n) {
        return 0;
    }
    for (int i = 0; i < n; i++) {
        if (x[i] != '0' && x[i] != '1') {
            return 0;
        }
    }
    return 1;
}

void add_impl(struct node* set, const char* x, int at) {
    if (at == n) {
        return;
    }
    if (x[at] == '0') {
        if (set->zero == 0) {
            set->zero = create();
        }
        add_impl(set->zero, x, at + 1);
    } else {
        if (set->one == 0) {
            set->one = create();
        }
        add_impl(set->one, x, at + 1);
    }
}

void wa() {
    printf("WA\n");
    exit(0);
}

void add_element(char* x) {
    if (--w < 0 || compiled || !check(x)) {
        wa();
    }
    add_impl(set, x, 0);
}

int check_element(char* x) {
    if (--r < 0 || !compiled || !check(x)) {
        wa();
    }
    struct node* current = set;
    for (int i = 0; i < n; i++) {
        if (x[i] == '0') {
            if (current->zero == 0) {
                return 0;
            }
            current = current->zero;
        }
        else {
            if (current->one == 0) {
                return 0;
            }
            current = current->one;
        }
    }
    return 1;
}

void compile(struct node* set, int at, struct node* compiledSet, char* element, char* compiled) {
    if (at == n) {
        for (int i = 0; i < n; i++) {
            compiled[i] = element[get_p(i)];
        }
        add_impl(compiledSet, compiled, 0);
        return;
    }
    if (set->zero != 0) {
        element[at] = '0';
        compile(set->zero, at + 1, compiledSet, element, compiled);
    }
    if (set->one != 0) {
        element[at] = '1';
        compile(set->one, at + 1, compiledSet, element, compiled);
    }
}

void compile_set() {
    if (compiled) {
        wa();
    }
    compiled = 1;
    struct node* compiledSet = create();
    char* element = (char*)malloc(n + 1);
    element[n] = 0;
    char* compiled = (char*)malloc(n + 1);
    compiled[n] = 0;
    compile(set, 0, compiledSet, element, compiled);
    set = compiledSet;
}

File: 2016/ioi_day2/messy/java/grader.java
import java.io.IOException;
import java.io.InputStream;
import java.util.Arrays;
import java.util.HashSet;
import java.util.InputMismatchException;
import java.util.Set;

/**
 * @author egor@egork.net
 */
public class grader {
    private static Set<String> set = new HashSet<>();
    private static boolean compiled = false;
    private static int n;
    private static int[] p;
    private static int w;
    private static int r;

    public static void main(String[] args) {
        InputReader in = new InputReader(System.in);
        n = in.readInt();
        w = in.readInt();
        r = in.readInt();
        p = new int[n];
        for (int i = 0; i < n; i++) {
            p[i] = in.readInt();
        }
        int[] answer = new messy().restore_permutation(n, w, r);
        
        
        System.out.print(answer[0]);
        for (int i = 1; i < n; i++) {
            System.out.print(" " + answer[i]);
        }
        System.out.println();
    }
    
    private static void wa() {
        System.out.println("WA");
        System.exit(0);
    }

    public static void add_element(String x) {
        if (--w < 0 || compiled || !check(x)) {
            wa();
        }
        set.add(x);
    }

    private static boolean check(String x) {
        if (x.length() != n) {
            return false;
        }
        for (int i = 0; i < n; i++) {
            if (x.charAt(i) != '0' && x.charAt(i) != '1') {
                return false;
            }
        }
        return true;
    }

    public static boolean check_element(String x) {
        if (--r < 0 || !compiled || !check(x)) {
            wa();
        }
        return set.contains(x);
    }

    public static void compile_set() {
        if (compiled) {
            wa();
        }
        compiled = true;
        Set<String> compiledSet = new HashSet<>();
        char[] compiledElement = new char[n];
        for (String s : set) {
            for (int i = 0; i < n; i++) {
                compiledElement[i] = s.charAt(p[i]);
            }
            compiledSet.add(new String(compiledElement));
        }
        set = compiledSet;
    }
}

class InputReader {
    private InputStream stream;
    private byte[] buf = new byte[1024];
    private int curChar;
    private int numChars;

    public InputReader(InputStream stream) {
        this.stream = stream;
    }

    public int read() {
        if (numChars == -1) {
            throw new InputMismatchException();
        }
        if (curChar >= numChars) {
            curChar = 0;
            try {
                numChars = stream.read(buf);
            } catch (IOException e) {
                throw new InputMismatchException();
            }
            if (numChars <= 0) {
                return -1;
            }
        }
        return buf[curChar++];
    }

    public int readInt() {
        int c = read();
        while (isSpaceChar(c)) {
            c = read();
        }
        int sgn = 1;
        if (c == '-') {
            sgn = -1;
            c = read();
        }
        int res = 0;
        do {
            if (c < '0' || c > '9') {
                throw new InputMismatchException();
            }
            res *= 10;
            res += c - '0';
            c = read();
        } while (!isSpaceChar(c));
        return res * sgn;
    }

    public static boolean isSpaceChar(int c) {
        return c == ' ' || c == '\n' || c == '\r' || c == '\t' || c == -1;
    }
}
